File: pytq/benchmark.py
#!/usr/bin/env python3
"""
Benchmark script for pytq Python bindings
"""

import time
import json
import sys

try:
    import pytq
except ImportError:
    print("Error: pytq module not found. Build and install it first.")
    sys.exit(1)

def benchmark_query(name, expression, data, iterations=1000):
    """Run a benchmark for a specific query"""
    start = time.perf_counter()
    
    result_count = 0
    for _ in range(iterations):
        results = pytq.query(expression, data)
        result_count = len(results)
    
    end = time.perf_counter()
    avg_time_ms = (end - start) * 1000 / iterations
    
    return {
        'name': name,
        'time_ms': avg_time_ms,
        'result_count': result_count
    }

def main():
    print("pytq Python Bindings Benchmarks")
    print("================================\n")
    
    # Load test data
    with open('../tq-core/tests/data/sample.json', 'r') as f:
        data = json.load(f)
    
    # Run benchmarks
    benchmarks = [
        ("Simple field access", ".metadata", data),
        ("Nested field", ".metadata.count", data),
        ("Array iteration", ".users[]", data),
        ("Array field fanout", ".users[].email", data),
        ("Nested array access", ".users[].roles[]", data),
    ]
    
    results = []
    for name, expr, test_data in benchmarks:
        result = benchmark_query(name, expr, test_data)
        results.append(result)
    
    # Print results
    print(f"{'Query':<30} {'Time (ms)':>10} {'Results':>10}")
    print("-" * 52)
    
    for result in results:
        print(f"{result['name']:<30} {result['time_ms']:>10.4f} {result['result_count']:>10}")
    
    print("\n Benchmarks completed successfully")

if __name__ == '__main__':
    main()

File: tq-core/tests/benchmark.cpp
#include "tq/tq.hpp"
#include <iostream>
#include <fstream>
#include <chrono>
#include <vector>
#include <string>

struct BenchmarkResult {
    std::string name;
    double time_ms;
    size_t result_count;
};

std::string read_file(const std::string& filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open file: " + filename);
    }
    std::string content((std::istreambuf_iterator<char>(file)),
                        std::istreambuf_iterator<char>());
    return content;
}

BenchmarkResult benchmark_query(const std::string& name, 
                                const std::string& expr,
                                const std::string& data,
                                int iterations = 1000) {
    auto start = std::chrono::high_resolution_clock::now();
    
    size_t total_results = 0;
    for (int i = 0; i < iterations; ++i) {
        auto results = tq::query(expr, data);
        total_results = results.size();
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    return BenchmarkResult{
        name,
        duration.count() / 1000.0 / iterations,
        total_results
    };
}

int main() {
    std::cout << "TQ Query Engine Benchmarks\n";
    std::cout << "===========================\n\n";
    
    try {
        // Load test data
        std::string data = read_file("tests/data/sample.json");
        
        std::vector<BenchmarkResult> results;
        
        // Benchmark various queries
        results.push_back(benchmark_query("Simple field access", ".metadata", data));
        results.push_back(benchmark_query("Nested field", ".metadata.count", data));
        results.push_back(benchmark_query("Array iteration", ".users[]", data));
        results.push_back(benchmark_query("Array field fanout", ".users[].email", data));
        results.push_back(benchmark_query("Nested array access", ".users[].roles[]", data));
        
        // Print results
        std::cout << "Query                          Time (ms)    Results\n";
        std::cout << "----------------------------------------------------\n";
        
        for (const auto& result : results) {
            std::cout << std::left << std::setw(30) << result.name
                      << std::right << std::setw(10) << std::fixed 
                      << std::setprecision(4) << result.time_ms
                      << std::setw(10) << result.result_count << "\n";
        }
        
        std::cout << "\n Benchmarks completed successfully\n";
        return 0;
        
    } catch (const std::exception& e) {
        std::cerr << "Benchmark failed: " << e.what() << "\n";
        return 1;
    }
}
